core/src/test/java/org/infinispan/globalstate/AbstractGlobalStateRestartTest.java

```diff
@@ -0,0 +1,197 @@
+import java.util.Map;
+import java.util.Set;
+
+import static org.testng.AssertJUnit.*;
+
+public abstract class AbstractGlobalStateRestartTest extends MultipleCacheManagersTest {
+
+   protected abstract int getClusterSize();
+
+   public static int DATA_SIZE = 1000;
+
+   @Override
+   protected boolean cleanupAfterMethod() {
+      return true;
+   }
+
+   @Override
+   protected boolean cleanupAfterTest() {
+      return false;
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      TestingUtil.recursiveFileRemove(TestingUtil.tmpDirectory(this.getClass().getSimpleName()));
+      createStatefulCacheManagers(true, -1);
+   }
+
+   protected void createStatefulCacheManagers(boolean clear, int extraneousNodePosition) throws Throwable {
+      int totalNodes = getClusterSize() + ((extraneousNodePosition < 0) ? 0 : 1);
+      int node = 0;
+      for (int i = 0; i < totalNodes; i++) {
+         if (i == extraneousNodePosition) {
+            // Create one more node if needed in the requested position
+            createStatefulCacheManager(Character.toString('@'), true);
+         } else {
+            createStatefulCacheManager(Character.toString((char) ('A' + (node++))), clear);
+         }
+      }
+   }
+
+   private void createStatefulCacheManager(String id, boolean clear) {
+      String stateDirectory = TestingUtil.tmpDirectory(this.getClass().getSimpleName() + File.separator + id);
+      if (clear)
+         TestingUtil.recursiveFileRemove(stateDirectory);
+      GlobalConfigurationBuilder global = GlobalConfigurationBuilder.defaultClusteredBuilder();
+      global.globalState().enable().persistentLocation(stateDirectory);
+
+      ConfigurationBuilder config = new ConfigurationBuilder();
+      applyCacheManagerClusteringConfiguration(config);
+      config.persistence().addSingleFileStore().location(stateDirectory);
+      addClusterEnabledCacheManager(global, config);
+   }
+
+   protected abstract void applyCacheManagerClusteringConfiguration(ConfigurationBuilder config);
+
+   protected void shutdownAndRestart(int extraneousNodePosition) throws Throwable {
+      Map<JGroupsAddress, PersistentUUID> addressMappings = createInitialCluster();
+
+      ConsistentHash oldConsistentHash = cache(0).getAdvancedCache().getDistributionManager().getConsistentHash();
+
+      // Shutdown the cache cluster-wide
+      cache(0).shutdown();
+
+      TestingUtil.killCacheManagers(this.cacheManagers);
+
+      // We should have some data here
+      for (int i = 0; i < getClusterSize(); i++) {
+         checkStateDirNotEmpty(manager(i).getCacheManagerConfiguration().globalState().persistentLocation());
+      }
+      this.cacheManagers.clear();
+
+      // Simulate a restart
+      JGroupsAddressCache.flushAddressCaches();
+
+      // Recreate the cluster
+      createStatefulCacheManagers(false, extraneousNodePosition);
+
+      // Healthy cluster
+      switch (extraneousNodePosition) {
+         case -1: {
```

Comment:
These case code could be in the calling methods, which could in turn be moved to the base class.